### ISSUE TITLE
parsing formfactor: asn1: syntax error: truncated tag or length (PIV)

### ISSUE THREAD

**Issue by phiekl** [joemiller/yk-attest-verify#2]
I've encountered the following issue, using the current master (d77c47bf):
```
$ yk-attest-verify piv att.pem signer.pem

Attestation Policy Checks:
✖ Unable to parse attestation cert: parsing extension: parsing formfactor: asn1: syntax error: truncated tag or length
```
I've tried exporting the attestation certificate using both ykman 4.0.1 and yubico-piv-tool, tried with RSA2048 and ECCP256 keys, and compiled the program with golang 1.14 (debian buster backports) and 1.15 (debian bullseye), but it just ends up like this.

By removing the whole `case e.Id.Equal(extIDFormFactor)` block from pkg/piv/attestation.go, it executes successfully:
```
$ yk-attest-verify piv att.pem signer.pem
YubiKey PIV Attestation:
 - Key slot       : 9a
 - YubiKey Version: v5.2.7
 - Serial #       : <redacted>
 - Formfactor     : Unspecified
 - PIN Policy     : Once
 - Touch Policy   : Cached

Attestation Policy Checks:
✔ All policy checks OK
```

If you'd like to see the .pem files I can send it privately.

This is a very nice project btw, and the openpgp part works great, thanks!

**Comment by joemiller** [comment]
@phiekl thanks for reporting this! I suppose I'm not entirely surprised. During initial implementation I was a bit surprised about how different and loosely defined the specs were from Yubi between the two attestation formats, as well as differences between firmware versions. And even with about 6 different yubikeys from different vintages there were interesting differences between versions.

I'm interested in understanding and fixing this issue.

It looks like this PIV attestation was generated from v5.2.7 firmware. The closest I have is a 5C-Nano w/ v5.2.4, so I wonder what may be going on.

Yes, if you wouldn't mind sending me the attestation .pem's that would be great. I'm not sure the best way to send them. Any suggestions? My GPG keys are available on keybase if that helps - https://keybase.io/joemiller

**Comment by joemiller** [comment]
Also - What form factor was the attestation generated from? Perhaps there is a newer yubikey we need to account for in the switch statement.

https://github.com/joemiller/yk-attest-verify/blob/master/pkg/piv/attestation.go#L247-L257

**Comment by phiekl** [comment]
@joemiller Thanks for your reply! I've emailed you the pem files.

I'm using the YubiKey 5C NFC. While using the pgp part of yk-attest-verify it's identified like:
```
 - Formfactor     : USB-C Keychain
 ```

**Comment by joemiller** [comment]
@phiekl great, thank you. i will take a look this week

**Comment by joemiller** [comment]
My initial read on the situation:

- In my own test environment I have a PIV attestation .pem that lacks the form factor extension, so I was not testing for that extension.
- It appears that the PIV form factor extension does not use ASN.1 encoding and instead puts a raw byte in the data. So the fix that is working for me at the moment is similar to what I had to do for the firmware version extension which is also not ASN.1 encoded: https://github.com/joemiller/yk-attest-verify/blob/master/pkg/piv/attestation.go#L208-L214

your attestation cert parses correctly after I make a similar change:
```console
YubiKey PIV Attestation:
 - Key slot       : 9a
 - YubiKey Version: v5.2.7
 - Serial #       : <redacted>
 - Formfactor     : USB-C Keychain
 - PIN Policy     : Once
 - Touch Policy   : Cached
```

I'm going to keep poking at it but it feels like the fix here is to assume the form-factor extension is encoded as a raw byte instead of ASN.1 encoded.

**Comment by joemiller** [comment]
@phiekl Should be fixed in v0.0.3 https://github.com/joemiller/yk-attest-verify/releases
